Track hypertables used during process utility hook execution

pmwkaa opened the pull request:
This patch does refactoring necessary to support execution of DDL commands on remote servers.

Basically it extends cross module api with `ddl_command_start` function and pass `ProcessUtilityArgs` context to it. Additionally it shares hypertables cache pin, which now happens before `process_ddl_command_start` function call and released afterwards.

New variables `hypertables_list`, `subcmd_list` added to `ProcessUtilityArgs`. These fields are used to keep a list of found hypertables during Utility/DDL statement parsing. This information will be used to distinct distributed hypertables and forward DDL commands to any remote servers associated with them.

Added functionaly and test to track DROP INDEX call and ensure it is used against only one hypertable at a time.

cevian commented on src/process_utility.c:
```diff
@@ -778,23 +810,23 @@ process_reindex(Node *parsetree)
-			ht = ts_hypertable_cache_get_entry(hcache, IndexGetRelation(relid, true));
+			ht = ts_hypertable_cache_get_entry(args->hcache, IndexGetRelation(relid, true));
 
 			if (NULL != ht)
```
no need to `process_add_hypertable` here? Even if this is an error now, it may not be in the future...

cevian commented on src/process_utility.c:
```diff
@@ -2659,7 +2696,20 @@ timescaledb_ddl_command_start(
+
+	handled = process_ddl_command_start(&args);
+
+	if (ts_cm_functions->ddl_command_start)
```
So I am a bit skeptical of this hook. It happens /before/ the action itself happens on the frontend (and such actions will sometimes throw errors when running on the frontend). Are we sure we want it running before  `prev_ProcessUtility(&args)` below?

cevian commented on src/process_utility.c:
```diff
@@ -183,13 +170,24 @@ relation_not_only(RangeVar *rv)
+static void
+process_add_subcmd(ProcessUtilityArgs *args, void *datum)
+{
+	args->subcmd_list = lappend(args->subcmd_list, datum);
```
so this subcmd_list doesn't have any ordering guarantees with regard to `args->hypertable_list` do we need it to?

erimatnor commented on src/process_utility.c:
```diff
@@ -2659,7 +2696,20 @@ timescaledb_ddl_command_start(
+	if (ts_cm_functions->ddl_command_start)
+		ts_cm_functions->ddl_command_start(&args);
+
+	ts_cache_release(args.hcache);
```
This should not be released here since `args` is passed to `prev_ProcessUtility` below. Also agree with @cevian that we probably want to do _all_ processing here first before calling the hook.

erimatnor commented on src/process_utility.c:
```diff
@@ -183,13 +170,24 @@ relation_not_only(RangeVar *rv)
+}
+
+static void
+process_add_subcmd(ProcessUtilityArgs *args, void *datum)
```
I vould avoid calling the parameter `datum` here as there is a specific `Datum` type in PostgreSQL and the argument here is AFAIK typically not a datum. If not `void`, I believe the correct type would be `Node *`, since all commands should be a node type.

pmwkaa commented:
List of changes:

* Reworked using OID list instead of keeping hypertable pointers list and pinning hypertable cache before doing hook processing. Current hypertable cache pinning implementation is using subtransaction id for matching, one can change during the hook execution.

* For operations `ALTER TABLE SET SCHEMA` and `ALTER TABLE RENAME` had to call previous process utility hook to properly update hypertable information, otherwise it makes not working finding hypertable in cache in tsl hook.

* Still have to run the tsl hook before PostgreSQL hook, because otherwise it makes hypertable used in `DROP TABLE` not be seen in the hook. Also to note that this logic is not straightforward. There are several situation when we do force call of previous utility hook inside main hook procedures.

* Removed subcmd_list, because its been used only for `ALTER TABLE` and exactly same information can be accessed via stmt object.

* Added ddl_hook C and SQL test, which installs `tsl_ddl_command_start` hook and prints debug info about queries and hypertables its been processing.

pmwkaa commented on src/process_utility.c:
```diff
@@ -640,12 +652,50 @@ process_drop_hypertable_chunks(DropStmt *stmt)
+		if (NULL != ht)
+		{
+			if (list_length(stmt->objects) != 1)
+				elog(ERROR, "cannot drop a hypertable index along with other objects");
```
I can do that, but its actually the same way as in `DROP TABLE` that been there before. (see: process_drop_hypertable_chunks)

pmwkaa commented on src/process_utility.c:
```diff
@@ -834,13 +891,18 @@ process_rename_table(Cache *hcache, Oid relid, RenameStmt *stmt)
+	ts_hypertable_set_name(ht, stmt->newname);
+
+	process_add_hypertable(args, ht);
+	return true;
```
Thats a bit different here. Return value here says that is should be handled in `process_rename`. Since we are working under cache pin taken in `process_rename` we should call processUtility after cache release and then pass it down. It should crash on assert in `hypertable_cache_create_entry`, because relation and and one that in our schema will not match.

pmwkaa commented on src/process_utility.c:
```diff
@@ -979,44 +1048,45 @@ process_rename_constraint(Cache *hcache, Oid relid, RenameStmt *stmt)
 	{
 		case OBJECT_TABLE:
-			process_rename_table(hcache, relid, stmt);
+			handled = process_rename_table(args, hcache, relid, stmt);
```
Thats related to comment above. `ALTER SCHEMA RENAME` is not handled here, but probably could as well. I mean, we could track involved hypertables in `hypertable_rename_schema_name` but its not clear that we should do it now, for example we limit DROP TABLE and DROP INDEX to work with 1 object at a time right now.

pmwkaa commented on tsl/test/expected/ddl_hook.out:
```diff
@@ -0,0 +1,202 @@
+RETURN OLD;
+END
+$BODY$;
+NOTICE:  test_ddl_command_hook: 0 hypertables, query: CREATE OR REPLACE FUNCTION test_trigger()
```
This NOTICE message because CREATE FUNCTION in also handled in utility hook and it did not involved any hypertables. Or do you mean something else?

pmwkaa commented on tsl/test/expected/ddl_hook.out:
```diff
@@ -0,0 +1,202 @@
+FOR EACH ROW EXECUTE PROCEDURE test_trigger();
+NOTICE:  test_ddl_command_hook: htable
+DROP TRIGGER htable_trigger_test on htable;
+NOTICE:  test_ddl_command_hook: 0 hypertables, query: DROP TRIGGER htable_trigger_test on htable;
```
Yes, it should. I'll work it out how to do that

cevian commented on src/process_utility.c:
```diff
@@ -721,14 +801,20 @@ process_grant_and_revoke_role(ProcessUtilityArgs *args)
-			process_drop_hypertable_chunks(stmt);
+			process_drop_hypertable_chunks(args, stmt);
+			break;
+		case OBJECT_INDEX:
```
I think this actually has to go through `process_ddl_sql_drop`, and maybe use another hook. The current approach would miss all CASCADING drops.  I think that merging this in the current state is dangerous since it gives the impression that things drops are caught when they aren't.

cevian commented on src/process_utility.c:
```diff
@@ -1026,6 +1129,13 @@ process_rename(Node *parsetree)
 	ts_cache_release(hcache);
+
+	/* In case if command modifies a hypertable meta-data we need to
+	 * process it first here to sync up, before passing to tsl hook */
```
please make the comment. `In case the command modifies the relation's name we need to process the update here so that the relation's name and our catalog metadata is in-sync before the tsl hook is called`

cevian commented on src/process_utility.c:
```diff
@@ -834,13 +924,18 @@ process_rename_table(Cache *hcache, Oid relid, RenameStmt *stmt)
-	else
-		ts_hypertable_set_name(ht, stmt->newname);
+
+	ts_hypertable_set_name(ht, stmt->newname);
```
I think we can just move this to `process_ddl_command_end` and avoid the special logic here. But I'll leave it up to you. You can decide.

cevian commented on tsl/test/expected/ddl_hook.out:
```diff
@@ -0,0 +1,203 @@
+-- This file and its contents are licensed under the Timescale License.
```
I would like some tests for cascading drops in here.

pmwkaa commented:
**List of changes**

* Added `ddl_command_end` and `sql_drop` event hooks.

* New issue came up. Since we are not using real `ddl_command_start` event but overloading process utility hook we might easily be in situation when `ddl_command_end` or `sql_drop` trigger starts before our tsl version of `ddl_command_start` function. This happens if we execute the tsl hook after previous utility hook.

* `RENAME TABLE` and `SET SCHEMA` were required to call previous utility hook before tsl processing to sync up relation change. This doesn't work any more. Instead we just keep a track of hypertables oids involved in those operations and save them for further processing during `ddl_command_end` trigger.

* Hypertable tracking in `DROP` operations been removed. They should be tracked by tsl-side during `sql_drop` execution for now.

* There are some situations when hypertable might show up both in `ddl_command_start` and `sql_drop` hooks during `ALTER TABLE DROP ...`. This should be resolved on tsl side of execution as well.

* Added more tests and rebased.

cevian commented on src/process_utility.c:
```diff
@@ -640,12 +647,47 @@ process_drop_hypertable_chunks(DropStmt *stmt)
 }
 
+static void
+process_drop_hypertable_index(ProcessUtilityArgs *args, DropStmt *stmt)
```
nit: I'd like a comment of why this is needed (pretty much what you said in the chat).

erimatnor commented on tsl/test/expected/ddl_hook.out:
```diff
@@ -0,0 +1,317 @@
+NOTICE:  test_sql_drop: constraint: public.htable.temp_chk
+NOTICE:  test_sql_drop: index
+NOTICE:  test_sql_drop: trigger
+NOTICE:  test_sql_drop: index
```
Why no `ddl_command_end` here?

erimatnor replied:
add comment in test that `ddl_command_end` is not expected here currently

erimatnor commented on tsl/test/src/test_ddl_hook.c:
```diff
@@ -0,0 +1,177 @@
+
+	/*
+	 * Hypertable oid from those commands is available in hypertable_list but
+	 * cannot be resolved here until standart utility hook will synchronize new
```
standart -> standard

cevian commented on tsl/src/CMakeLists.txt:
```diff
@@ -28,6 +28,7 @@ set_target_properties(${TSL_LIBRARY_NAME} PROPERTIES
 
 target_compile_definitions(${TSL_LIBRARY_NAME} PUBLIC TS_TSL)
 target_compile_definitions(${TSL_LIBRARY_NAME} PUBLIC TS_SUBMODULE)
+target_compile_definitions(${TSL_TESTS_LIB_NAME} PUBLIC TS_SUBMODULE)
```
I think this should go into tsl/test/src/CMakeLists if possible

pmwkaa commented:
Changes:

* added `DROP TABLE` and `DROP INDEX` to actual `ddl_command_end` trigger in `ddl_triggers.sql`
* looks like that our current `CREATE INDEX` implementation function does not call origin standard hook and because of that `ddl_command_end` not runs. Not sure how and that we need to fix it in this PR.
* made refactoring and typo fixes
